library/src/main/java/com/mux/player/media/MuxDataSource.kt

```diff
@@ -98,4 +101,209 @@ class MuxDataSource private constructor(
     cacheWriter?.close()
     upstream?.close()
   }
+
+  private fun openAndInitRevalidating(dataSpec: DataSpec, readHandle: ReadHandle): Long {
+    val revalidateRequestHeaders = mutableMapOf<String, String>()
+    revalidateRequestHeaders.putAll(dataSpec.httpRequestHeaders)
+    revalidateRequestHeaders["If-None-Match"] = readHandle.fileRecord.etag
+
+    val revalidateSpec = dataSpec.buildUpon()
+      .setHttpMethod(DataSpec.HTTP_METHOD_GET) // deviate from spec to save the round-trip w/cdn
+      .setHttpRequestHeaders(revalidateRequestHeaders)
+      .build()
+
+    val upstreamBytes = openAndInitFromRemote(revalidateSpec, RevalidatingDataSource.Factory())
+    val upstream = this.upstream!! // set by initAndOpenUpstream
+
+    Log.d(TAG, "revalidation: HTTP ${upstream.responseCode}. $upstreamBytes available")
+    return if (upstream.responseCode != 304) {
+      // Entry wasn't valid anymore, but we did a GET so the body's ready to read and we're done
+      Log.d(TAG, "revalidation: Entry was NOT valid, getting from network")
+
+      // todo - we *could* delete the row here, but consider that stale items can be used if
+      //  state-while-revalidate or stale-while-error or if we're disconnected (unless must-revalidate)..
+      //  Right now, we saved time by assuming state-while-error and *not* must-revalidate so we
+      //    would just wanna keep the stale entries around in case we can use them
+
+      upstreamBytes
+    } else {
+      Log.d(TAG, "revalidation: Entry WAS still valid, getting from cache")
+      // Entry was still valid, so read from cache instead
+      upstream.close()
+      this.upstream = null
+
+      // todo -
```

Comment:
issue(non-blocking): we should update the cache directive values here if they're updated. e.g. we should get a fresh max-age so we can avoid revalidating. Not a blocker but would be great to add before merging.

Reply:
wait no let's fast-follow!